common/konfig.py

```diff
@@ -0,0 +1,447 @@
+        self['snapshots.ssh.check_ping'] = value
+
+
+class Konfig(metaclass=singleton.Singleton):
+    """Manage configuration data for Back In Time.
+
+    Dev note:
+
+        That class is a replacement for the `config.Config` class.
+    """
+
+    _DEFAULT_VALUES = {
+        'global.hash_collision': 0,
+        'global.language': '',
+        'global.use_flock': False,
+    }
+
+    _DEFAULT_SECTION = '[bit]'
+
+    def __init__(self, config_path: Path = None):
+        if not config_path:
+            xdg_config = os.environ.get('XDG_CONFIG_HOME',
+                                        os.environ['HOME'] + '/.config')
+            self._path = Path(xdg_config) / 'backintime' / 'config'
+        else:
+            self._path = config_path
+
+        logger.debug(f'Config path used: {self._path}')
+
+        self.load()
+
+        # Names and IDs of profiles
+        # EXtract all relevant lines of format 'profile*.name=*'
+        name_items = filter(
+            lambda val:
+                val[0].startswith('profile') and val[0].endswith('name'),
+            self._conf.items()
+        )
+        self._profiles = {
+            name: int(pid.replace('profile', '').replace('.name', ''))
+            for pid, name in name_items
+        }
+
+    def __getitem__(self, key: str) -> Any:
+        try:
+            return self._conf[key]
+        except KeyError:
+            return self._DEFAULT_VALUES[key]
+
+    def __setitem__(self, key: str, val: Any) -> None:
+        self._conf[key] = val
+
+    def profile(self, name_or_id: Union[str, int]) -> Profile:
+        if isinstance(name_or_id, int):
+            profile_id = name_or_id
+        else:
+            profile_id = self._profiles[name_or_id]
+
+        return Profile(profile_id=profile_id, config=self)
+
+    @property
+    def profile_names(self) -> list[str]:
+        return list(self._profiles.keys())
+
+    @property
+    def profile_ids(self) -> list[int]:
+        return list(self._profiles.values())
+
+    def load(self):
+        """Load configuration from file like object."""
+        @contextlib.contextmanager
+        def _path_or_buffer(path_or_buffer: Union[Path, StringIO]
+                            ) -> Union[TextIOWrapper, StringIO]:
+            """Using a path or a in-memory file (buffer) with a with
+            statement."""
+            try:
+                # It is a regular file
+                path_or_buffer = path_or_buffer.open('r', encoding='utf-8')
+                print(f'{type(path_or_buffer)=}')
+
+            except AttributeError:
+                # Assuming a StringIO instance as in-memory file
+                pass
+
+            yield path_or_buffer
+
+            try:
+                # regular file: close it
+                path_or_buffer.close()
+
+            except AttributeError:
+                # in-memory file: "cursor" back to first byte
+                path_or_buffer.seek(0)
```

Comment:
Mhm... I am not sure but see no need for it.
To my understanding of context managers that block you enclosded in "finally" will be called every time no matter what happens. This is because I do use that context manager with `with...` statement. The code after the `yield` line is executed under all circumstance when the `with` block is left.

I might be wrong.

Reply:
I had never encountered this magic context manager until I saw it here. There was a line in the [docs](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager) which made me concerned about what would happen in the case of an error:

> If an unhandled exception occurs in the block, it is _reraised inside the generator_ at the point where the yield occurred. Thus, you can use a [try](https://docs.python.org/3/reference/compound_stmts.html#try)…[except](https://docs.python.org/3/reference/compound_stmts.html#except)…[finally](https://docs.python.org/3/reference/compound_stmts.html#finally) statement to trap the error (if any), or ensure that some cleanup takes place.

Here is a test:

``` python
@contextmanager
def manager():
    try:
        print("The game begins...")
    except Error as err:
        print("Oh no, an error!")
        print(err)

    yield "thing"

    try:
        print("The game ends.")
    except Error as err:
        print("Oh no, another error!")
        print(err)

with manager() as m:
    print("Let's break maths:", 1/0)
```

If you run this, you will see that the cleanup code is never run.
It never prints `"The game ends."`:

``` python
The game begins...
---------------------------------------------------------------------------
ZeroDivisionError                         Traceback (most recent call last)
Cell In[52], line 2
      1 with manager() as m:
----> 2     print("Let's break maths:", 1/0)

ZeroDivisionError: division by zero
```

So, I think we need the `finally:` block.

Here's another example which ensures that `yield "thing"` is safely wrapped in a `try: finally:`:

``` python
@contextmanager
def manager():
    try:
        try:
            print("The game begins...")
        except Error as err:
            print("Oh no, an error!")
            print(err)

        yield "thing"

    finally:
        try:
            print("The game ends.")
        except Error as err:
            print("Oh no, another error!")
            print(err)
```

If you run this, you will see that the cleanup runs successfully, and `"The game ends."` is successfully printed before the error is raised:

``` python
The game begins...
The game ends.
---------------------------------------------------------------------------
ZeroDivisionError                         Traceback (most recent call last)
Cell In[54], line 2
      1 with manager() as m:
----> 2     print("Let's break maths:", 1/0)

ZeroDivisionError: division by zero
```

Reply:
Mhm... reminds me of something I read in the past but never understood. 🤣 Now I have real world illustration. 🎉 Thank you for this.

Anyway, I don't think the context manager is a good solution at all. There is a rule about testing: Never modify productive code just because of a unit test. One alterative to the context manager would be a virtual file system via `pyfakefs` while running the test suite. I really like PyFakeFS for specific cases but in our case it would make things complicated without much benefit.

I might remove the file handling logic from `Konfig` class. It should be something like this:

    fp = Path.home() / 'somewhere' / 'configfile'
    with fp.open('r', encoding='utf-8') as handle:
        config = Konfig(handle)

        # or ...
        config = Konfig()
        config.load(handle)

This way it is more easier using in-memory files in a unit test:

    def test_foobar():
        buffer = StringIO('filed=value')
        config = Konfig()
        config.load(handle)

Reply:
I think that's a good call.